pkg/flatrpc/conn.go

```diff
@@ -30,36 +32,54 @@ type Serv struct {
 	ln   net.Listener
 }
 
-func ListenAndServe(addr string, handler func(*Conn)) (*Serv, error) {
+func Listen(addr string) (*Serv, error) {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		return nil, err
 	}
+	return &Serv{
+		Addr: ln.Addr().(*net.TCPAddr),
+		ln:   ln,
+	}, nil
+}
+
+// Serve accepts incoming connections and calls handler for each of them.
+// An error returned from the handler stops the server and aborts the whole processing.
+func (s *Serv) Serve(baseCtx context.Context, handler func(context.Context, *Conn) error) error {
+	eg, ctx := errgroup.WithContext(baseCtx)
 	go func() {
-		for {
-			conn, err := ln.Accept()
+		// If the context is cancelled, stop the server.
+		<-ctx.Done()
+		s.Close()
+	}()
+	for {
+		conn, err := s.ln.Accept()
+		if err != nil && errors.Is(err, net.ErrClosed) {
+			break
+		}
+		eg.Go(func() error {
 			if err != nil {
-				if errors.Is(err, net.ErrClosed) {
-					break
-				}
 				var netErr *net.OpError
 				if errors.As(err, &netErr) && !netErr.Temporary() {
-					log.Fatalf("flatrpc: failed to accept: %v", err)
+					return fmt.Errorf("flatrpc: failed to accept: %w", err)
 				}
 				log.Logf(0, "flatrpc: failed to accept: %v", err)
-				continue
+				return nil
 			}
+
+			connCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			c := NewConn(conn)
+			// Closing the server does not automatically close all the connections.
 			go func() {
-				c := NewConn(conn)
-				defer c.Close()
-				handler(c)
+				<-connCtx.Done()
```

Comment:
These goroutines will be effectively leaking. In the manager we will have them piling up infinity.

Reply:
Why will they be leaking?
The `connCtx` is cancelled after the `hander` exits - i.e. after the corresponding VM stopped. After that, the goroutine will be unblocked and exit.

Reply:
Right. I thought it's the parent server context.